src/java.base/share/classes/java/lang/String.java

```diff
@@ -522,13 +522,13 @@ public String(int[] codePoints, int offset, int count) {
             return;
         }
         if (COMPACT_STRINGS) {
-            byte[] val = StringLatin1.toBytes(codePoints, offset, count);
-            if (val != null) {
-                this.coder = LATIN1;
-                this.value = val;
-                return;
+            byte[] val = StringUTF16.compress(codePoints, offset, count);
+            this.coder = StringUTF16.coderFromArrayLen(val, count);
+            if (UTF16 == coder) {
+                initCompressionFlag();
             }
-            initCompressionFlag();
+            this.value = val;
```

Comment:
Suggest reordering places that call `initCompressionFlag()` and using `coder == UTF16` consistently (see line 874).
```java
            this.value = val;
            if (coder == UTF16) {
                initCompressionFlag();
            }
```
Written this way makes the JIT's job a little easier.

While you're here, please remove the trailing tabs from line 1617.